### PR TITLE
SentinelOne Phantom app enhancement v2.2.0

### PR DESCRIPTION
**Pull request by abhijeeta-metron**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [X] The readme.html has been reviewed and added / updated if needed (for bug fixes / features)
- [X] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [X] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [X] Verify all checks are passing.
- [X] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [X] Do NOT submit updates to dependencies unless it fixes an issue.
- [X] Limit your pull request to one application and submit multiple pull requests if necessary.
#### Please check the type of change your PR introduces:
- [X] New App
- [X] Bugfix
- [X] Feature
- [X] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no api changes)
- [X] Documentation
- [X] Other (please describe): 

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describes high-level points about the changes for the upcoming GA release.
-  The following actions are modified so that they work with IP addresses too:
    1. Quarantine device: This command isolate (quarantine) endpoints from the network.
    2. Unquarantine device: This command reconnects to the network at all endpoints.
    3. Scan endpoint: This command runs a Full Disk Scan on Agents that match the filter.
    4. Get endpoint info: Get detailed information about an endpoint/agent.
-  Modified command to support new SentinelOne API:
    1. Block hash: Add a file hash to the global blocklist.
-  The following new actions are implemented:
    1. AGENT ACTIONS
       -  Abort Scan: Stop a Full Disk Scan on endpoint/agent.
       -  Broadcast Message: Send a message through the Agents that users can see.
       -  Fetch Files (P1): Fetch files from endpoints to analyze the root of threats.
       -  Fetch Firewall Logs: Get Firewall Control events in the local log file.
       -  Fetch Firewall Rules: Fetch firewall rules.
       -  Shutdown: Shutdown endpoint.
    2.  APPLICATION RISK
        -  Get Applications: Get the applications, and their data, installed on endpoints.
        -  Get CVEs: Get known CVEs for applications that are installed on endpoints with Application Risk-enabled Agents.
    3. DEVICE CONTROL
       -  Get Device Control Events (given a hostname or IP): Get the data of Device Control events on Windows and macOS 
       endpoints.
   4. FIREWALL CONTROL
      -  Get Firewall Rules: Get the Firewall Control rules for a scope specified.
      -  Create Firewall Rule: Create a Firewall Control rule.
   5. HASHES
       - Hash Reputation: Get the reputation of a hash, given the required SHA1.
   6. THREAT NOTES:
       -  Add Note to Multiple: Add a threat note to multiple threats.
       - Get Threat Notes: Get the threat notes.
   7. THREATS:
      - Download from cloud: Download threat file from cloud.
      - Export Mitigation Report: Export the mitigation report of threat.
      - Export Threat Timeline: Export a threat's timeline.
      - Export Threats: Export data of threats.
      - Fetch Threat File: Fetch a file associated with the threat.
      - Update Threat Analyst Verdict: Change the verdict of a threat, as determined by a Console user.
      - Get Threat Timeline: Get a threat's timeline.
      - Update Threat Incident: Update the incident details of a threat.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

**Comment by abhijeeta-metron** [comment]
Need clarification for the "Static Tests" error.

**Comment by mattsayar-splunk** [comment]
Looks like the final piece of the puzzle revolves around the actions names (seen in the static test and pasted below). We try to maintain the same action names between all the apps as much as we can, but sometimes it's not entirely possible. Can you please review the names of the actions and let us know if it's possible to change any of them to something that already exists?

"Action `abort scan` not known. Would `url scan` be better?",
                        "Action `shutdown endpoint` not known.",
                        "Action `broadcast message` not known.",
                        "Action `fetch files` not known. Would `get file` be better?",
                        "Action `fetch firewall rules` not known.",
                        "Action `fetch firewall logs` not known.",
                        "Action `get applications` not known. Would `list applications` be better?",
                        "Action `get cves` not known. Would `get acls` be better?",
                        "Action `get devicecontrol events` not known.",
                        "Action `get firewall rules` not known. Would `list firewall rules` be better?",
                        "Action `create firewall rule` not known. Would `delete firewall rule` be better?",
                        "Action `hash reputation` not known. Would `host reputation` be better?",
                        "Action `get threat notes` not known.",
                        "Action `add threat note` not known.",
                        "Action `export threat timeline` not known.",
                        "Action `export mitigation report` not known.",
                        "Action `export threats` not known.",
                        "Action `fetch threat file` not known.",
                        "Action `update threat analystverdict` not known.",
                        "Action `get threat timeline` not known.",
                        "Action `update threat incident` not known.",
                        "Action `download from cloud` not known."

**Comment by abhijeeta-metron** [comment]
@mattsayar-splunk we are getting one error on FIPS compliance. The app does not store any credentials, can we mark it as fips_compliant = true? We don't see the key in any of the other apps.

**Comment by mattsayar-splunk** [comment]
@abhijeeta-metron The FIPS check is a new feature flag. If you aren't sure of your code's FIPS compliance, it's safe to just declare it "false" and ignore the error. We'll remove this check in the near future

**Comment by abhijeeta-metron** [comment]
@mattsayar-splunk We have resolved all the errors except these static tests errors we cant resolve.

**Comment by haritas-crest** [comment]
Hi @abhijeeta-metron, can you please provide me the access to the repo(metron-labs:sentinelone-phantom-2.2.0)?

**Comment by abhijeeta-metron** [comment]
@haritas-crest we have sent an invite for repo access.

**Review by rgil-splunk — APPROVED** [review]

**Review by ricardo-phantom — APPROVED** [review]

File: sentinelone_connector.py [hunk 1]
```diff
@@ -957,7 +979,9 @@ def _handle_export_threats(self, param):
                     return action_result.set_status(phantom.APP_ERROR, "Did not get proper response from the server")
                 self.save_progress('Agent query: {}'.format(ret_val))
 
-                if ret_val == '0':
+                if ret_val == '-1':
```
Comment (haritas-crest) [inline sentinelone_connector.py:982]: Hi @abhijeeta-metron,
While testing for the negative scenarios, we observed that the 'export threats' action was passing with invalid parameter values. To handle this case, we have added a check for the ‘-1’ string value after the get_agent_id method call. Similarly, we have added the check for '-1' string in other actions as well. 

Action result status before adding -1 check:
image

Action result status after adding -1 check:
image

Please let us know if you have any queries regarding this change.

Comment (haritas-crest) [inline sentinelone_connector.py:982]: Hi @abhijeeta-metron ,
While testing for the negative scenarios, we observed that the 'export threats' action was passing with invalid parameter values. To handle this case, we have added a check for the ‘-1’ string value after the get_agent_id method call. Similarly, we have added the check for '-1' string in other actions as well.

Action result status before adding -1 check:
image

Action result status after adding -1 check:
image

Please let us know if you have any queries regarding this change.

**Comment by haritas-crest** [comment]
Hi @abhijeeta-metron 
While testing for the negative scenarios, we observed that the 'export threats' action was passing with invalid parameter values. To handle this case, we have added a check for the ‘-1’ string value after the get_agent_id method call. Similarly, we have added the check for '-1' string in other actions as well. 
image


Action result status before adding -1 check:
image

Action result status after adding -1 check:
image

Please let us know if you have any queries regarding this change.

**Comment by abhijeeta-metron** [comment]
@haritas-crest Sure, I will check. Thanks

**Comment by abhijeeta-metron** [comment]
@haritas-crest we are ok with the changes

File: readme.html [hunk 2]
```diff
@@ -1,21 +1,69 @@
...
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+  either express or implied. See the License for the specific language governing permissions
+  and limitations under the License.
 -->
 
 <html>
   <head></head>
   <body>
-    This version of the SentinelOne app is compatible with Phantom <b>4.9.39220+</b>
+    This version of the SentinelOne app is compatible with Splunk SOAR version <b>5.0.0</b> and above.
```
Comment (mjiyani-crest) [inline readme.html]:
```suggestion
    This version of the SentinelOne app is compatible with Splunk SOAR version <b>5.1.0</b> and above.
```

File: sentinelone_connector.py [hunk 3]
```diff
@@ -417,29 +687,587 @@ def _handle_get_threat_info(self, param):
...
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        ip_hostname = param.get('ip_hostname')
+        params = None
+        if ip_hostname:
+            try:
+                ret_val = self._get_computer_name(ip_hostname, action_result)
+            except Exception:
+                return action_result.set_status(phantom.APP_ERROR, "Did not get proper response from the server")
+            self.save_progress('Agent query: {}'.format(ret_val))
+
```
Comment (mjiyani-crest) [inline sentinelone_connector.py:711]: -1 handling is required or not? Pls check and do the needfull.

File: readme.html (hunk 2 above)
Comment (haritas-crest) [reply readme.html]: Fixed

File: sentinelone_connector.py (hunk 3 above)
Comment (haritas-crest) [reply sentinelone_connector.py:711]: Done

**Pull request merged**